### PR TITLE
Fix CI pipeline errors and add model validation.

### PR DESCRIPTION
**Pull request by GhaithChaabane**
This PR fixes the CI pipeline failure in the build-and-push-base, linting and unit test steps by updating pip install commands, notebook linting paths, pytest invocation, and the base Dockerfile.
1- fix CI pipeline errors base image build, linting, and testing.
Changes include the CI Workflow in the build-and-test.yml file by :
- Updated pip commands to use **--no-cache-dir** instead of -**-no-deps/ --ignore-installed** to prevent dependency conflicts. 
 - Linting path updated to target only demo notebooks: from nbqa pylint notebooksnbqa  which was linting old kso (kso_utils) to **pylint notebooks/kso_demo_notebooks**
  
- Test commands updated to install pytest correctly and target the **test/** directory. **pytest -vv test/**
- update the toml file and requirements.txt to insure that the package **mlflow-export-import** is bing installed from the git commit containing the latest updated parameters, rather than the older version available via pip

- in **base/Dockerfile** i copied the full repo and set working directory for consistent pip installs and added **--break-system-packages** to avoid CI build failures.

2- adding model valadation
-for the Notebooks i added model valadation by adding **model_validation** methode that takes **project** and **data_path** parameters, the user can validate the model either after training in the training notebook or if they wish directly in the model inference notebook to validate a registered model.

### CODE REVIEW

**Comment by trossi** [comment]
@GhaithChaabane I opened #508 that fixes the CI issues in this PR.

File: notebooks/kso_demo_notebooks/kso2_demo_Model_Inference.ipynb [hunk 1]
```diff
@@ -48,6 +48,18 @@
...
   },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "0e97165e",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "image_path = (\n",
+    "    \"~/desktop/kso/kso/projects/test_project_1/coco8/images/val/000000000036.jpg\"\n",
```
[the remark this answers is not in the document]
Comment (GhaithChaabane) [reply notebooks/kso_demo_notebooks/kso2_demo_Model_Inference.ipynb]: That path was only an example to illustrate the parameters.

File: src/kso/trainer.py [hunk 2]
```diff
@@ -304,3 +307,55 @@ def internal_model(self, model: PyFuncModel):
...
+        run_name = run.info.run_name
+        val_run_name = f"validation_{run_name}"
+
+        uri_path = Path(urlparse(uri).path)
+        model_path = os.path.join(uri_path, "artifacts/best.pt")
+        yolo_model = YOLO(model_path)
+
+        with mlflow.start_run(experiment_id=experiment_id, run_name=val_run_name):
+
+            results = yolo_model.val(
```
Comment (louisrf) [inline src/kso/trainer.py:365]: I think "Split" should be a parameter, not hardcoded to "val". Val metrics are optimistically biased as the validation set influences which checkpoint gets saved as best.pt. For publishable, unbiased evaluation the default should be "test". Could you change it to something like

def model_validation(
    self,
    project: Project,
    model: PyFuncModel = None,
    data_path: str = None,
    split: str = "test",
    imgsz=640,
    batch=8,
):

and then passing it through

results = yolo_model.val(
    data=str(data_path),
    split=split,
    imgsz=imgsz,
    batch=batch,
)

File: src/kso/trainer.py [hunk 3]
```diff
@@ -304,3 +307,55 @@ def internal_model(self, model: PyFuncModel):
         if not isinstance(pytorch_model, nn.Module):
             raise TypeError(f"model {pytorch_model} is not a pytorch model")
         return pytorch_model
+
+    def model_validation(
+        self,
+        project: Project,
+        model: PyFuncModel = None,
+        data_path: str = None,
+        imgsz=1000,
```
Comment (louisrf) [inline src/kso/trainer.py]: imgsz=1000 isn't generally the standard, YOLO training defaults to 640 and then if you evaluate at a different resolution than training the metrics aren't really comparable. Should default to 640 for consistency, or even better, if it can be read from the training run's logged parameters

File: src/kso/trainer.py [hunk 4]
```diff
@@ -304,3 +307,55 @@ def internal_model(self, model: PyFuncModel):
...
+        # Get the run object
+        run = client.get_run(model.metadata.run_id)
+        # Get the experiment ID
+        experiment_id = run.info.experiment_id
+        run_name = run.info.run_name
+        val_run_name = f"validation_{run_name}"
+
+        uri_path = Path(urlparse(uri).path)
+        model_path = os.path.join(uri_path, "artifacts/best.pt")
+        yolo_model = YOLO(model_path)
```
Comment (louisrf) [inline src/kso/trainer.py:361]: Optional existence check if the path doesn't exist or gets nested differently so you don't get any unclear errors

uri_path = Path(urlparse(uri).path)
model_path = Path(uri_path) / "artifacts" / "best.pt"
if not model_path.exists():
    raise FileNotFoundError(f"Model weights not found: {model_path}")
yolo_model = YOLO(str(model_path))

**Review by louisrf — COMMENTED** [review]
On notebook placement: I'd remove the model_validation call from kso2_demo_Model_Inference.ipynb and keep it only in kso2_demo_Model_Training.ipynb.

They're fundamentally different operations for our users, which I think will just confuse them about what the notebook is for. If the use case is "validate a model I trained in a previous session,", which i'd say in my case is rare that should still be a training/evaluation workflow, and the user could just load existing weights and run the eval in Notebook 02, they'd just need to load the project/model and select the version number they want during that operation.

Also the MLflow integration for logging validation metrics as a separate run is a nice pattern, will be very helpful

File: src/kso/trainer.py (hunk 3 above)
Comment (GhaithChaabane) [reply src/kso/trainer.py]: For now i’ve set imgsz=640 as a temporary value but once we implement preprocessing for the input dataset we can either read the image size from the logged parameters of the training run or retain the default value.

File: src/kso/trainer.py (hunk 2 above)
Comment (GhaithChaabane) [reply src/kso/trainer.py:365]: Removed model_validation from kso2_demo_Model_Inference.ipynb and kept the training notebook version modified to handle only-registered models for now.

**Review by louisrf — APPROVED** [review]
